Add thread sanitizer run to daily CI

skaslev opened the pull request:
Add thread sanitizer run to daily CI.

Few tests are skipped in tsan runs for two reasons:
 * Stack trace producing tests (oom, `unit/moduleapi/crash`, etc) are tagged `tsan:skip` because redis calls `backtrace()` in signal handler which turns out to be signal-unsafe since it might allocate memory (e.g. glibc 2.39 does it through a call to `_dl_map_object_deps()`).
 * Few tests become flaky with thread sanitizer builds and don't finish in expected deadlines because of the additional tsan overhead. Instead of skipping those tests, this can improved in the future by allowing more iterations when waiting for tsan builds.

Deadlock detection is disabled for now because of tsan limitation where max 64 locks can be taken at once.

There is one outstanding (false-positive?) race in jemalloc which is suppressed in `tsan.sup`.

Fix few races thread sanitizer reported having to do with writes from signal handlers. Since in multi-threaded setting signal handlers might be called on any thread (modulo pthread_sigmask) while the main thread is running, `volatile sig_atomic_t` type is not sufficient and atomics are used instead.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/redis/redis?pullRequest=13964) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/redis/redis?pullRequest=13964) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/redis/redis?pullRequest=13964) it.</sub>

sundb commented:
@skaslev there are still lots of false positive report from thread ASAN, how can we tackle it?

ShooterIT commented:
maybe we can use `REDIS_NO_SANITIZE("thread")` to suppress if false positive reports are about function

skaslev commented:
@sundb I tried addressing all the issues tsan complained about and had to disable jemalloc for tsan daily ci runs since the clang build crashes immediately.

There are still several tests that flake occasionally:
```
2025-04-29T12:56:47.6422636Z [err]: stats: eventloop metrics in tests/unit/info.tcl
2025-04-29T12:56:47.6435509Z Expected '3776230' to be less than '3749344' (context: type eval line 19 cmd {assert_lessthan $el_sum2 [expr $el_sum1+100000] } proc ::test)

2025-04-29T12:57:59.7379989Z [err]: Active Expire - deletes hash that all its fields got expired (listpackex) in tests/unit/type/hash-field-expire.tcl
2025-04-29T12:57:59.7381354Z Expected '1' to be equal to '0' (context: type eval line 10 cmd {assert_equal 1 [r HEXISTS hrand$h f$i]} proc ::test)

2025-04-29T12:58:23.0980365Z [err]: HINCRBY - preserve expiration time of the field in tests/unit/type/hash-field-expire.tcl
2025-04-29T12:58:23.0982962Z Expected '-2' to be between to '1' and '20' (context: type eval line 6 cmd {assert_range [r HPTTL h1 FIELDS 1 f1] 1 20} proc ::test)

2025-04-29T12:58:25.2050117Z [err]: Active Expire - deletes hash that all its fields got expired (hashtable) in tests/unit/type/hash-field-expire.tcl
2025-04-29T12:58:25.2051431Z Expected '1' to be equal to '0' (context: type eval line 10 cmd {assert_equal 1 [r HEXISTS hrand$h f$i]} proc ::test)


2025-04-29T12:59:50.8033087Z [err]: BLPOP unblock but the key is expired and then block again - reprocessing command in tests/unit/type/list.tcl
2025-04-29T12:59:50.8037817Z Expected '1217' to be between to '1000' and '1150' (context: type eval line 23 cmd {assert_range [expr $end-$start] 1000 1150} proc ::test)

2025-04-29T13:00:45.7264885Z [err]: Consumer seen-time and active-time in tests/unit/type/stream-cgroups.tcl
2025-04-29T13:00:45.7269668Z Expected [dict get name Alice pending 1 idle 119 inactive 221 idle] < 80 (context: type eval line 23 cmd {assert {[dict get $consumer_info idle] < 80} } proc ::test)

2025-04-29T12:56:05.4372185Z [err]: Active Expire - deletes hash that all its fields got expired (listpackex) in tests/unit/type/hash-field-expire.tcl
2025-04-29T12:56:05.4383811Z Expected '1' to be equal to '0' (context: type eval line 10 cmd {assert_equal 1 [r HEXISTS hrand$h f$i]} proc ::test)

2025-04-29T12:56:28.0874030Z [err]: Active Expire - deletes hash that all its fields got expired (hashtable) in tests/unit/type/hash-field-expire.tcl
2025-04-29T12:56:28.0931971Z Expected '1' to be equal to '0' (context: type eval line 15 cmd {assert_equal 1 [r HEXISTS same$h f$i]} proc ::test)

2025-04-29T13:06:32.1219902Z [err]: Test FLUSHALL aborts bgsave in tests/integration/rdb.tcl
2025-04-29T13:06:32.1220646Z bgsave not aborted
```

tezc commented on .github/workflows/daily.yml:
```diff
@@ -377,7 +377,7 @@ jobs:
       run: |
         sudo apt-get install vmtouch
-        mkdir /tmp/master 
+        mkdir /tmp/master
```
Maybe later sometime but at some point please cleanup unnecessary changes. It pollutes blame log and makes it harder to cherry-pick the commit for fork maintainers.

skaslev replied:
Reverted unnecessary changes.

tezc commented on .github/workflows/daily.yml:
```diff
@@ -653,6 +653,53 @@ jobs:
+          sudo apt-get install tcl8.6 tclx -y
+      - name: test
+        if: true && !contains(github.event.inputs.skiptests, 'redis')
+        run: HAS_TSAN=1 TSAN_OPTIONS="detect_deadlocks=0,suppressions=tsan.supp" ./runtest --config io-threads 4 --accurate --verbose --dump-logs ${{github.event.inputs.test_args}}
```
I think we should try to run it without `io-thread` config first. I hope we'll see fewer failures on single thread. Once we clear those, we can try with io-thread config.

sundb commented:
```
2025-04-29T13:06:32.1219902Z [err]: Test FLUSHALL aborts bgsave in tests/integration/rdb.tcl
2025-04-29T13:06:32.1220646Z bgsave not aborted
```

@skaslev This is because SANITIZER intercepts signal processing, preventing us from receiving the signal indicating the end of the rdb.
ref: https://github.com/google/sanitizers/issues/1179#issuecomment-995414409

skaslev commented:
> @skaslev are these tests flaky or failing all the time? If it is flaky, then most probably it is due to timing, then I assume we can fix those one way or another. Btw, if I understand correctly, certain test files are causing extreme pain, maybe we can skip those if we can't fix those. e.g. corrupt-fuzzer.tcl (via --skipfile option).

They're flaky and pass locally but occasionally fail. Probably because tsan adds some overhead and sometimes we miss the deadlines some of tests expect.

tezc commented:
> FATAL: ThreadSanitizer: unexpected memory mapping 0x5fe9fadd1000-0x5fe9fae47000

I get this error on my local when I start a Redis instance. Looks like this is the issue: https://github.com/google/sanitizers/issues/1716 
There is a suggested workaround but not sure if I want to do it.

sundb commented:
> > FATAL: ThreadSanitizer: unexpected memory mapping 0x5fe9fadd1000-0x5fe9fae47000
> 
> I get this error on my local when I start a Redis instance. Looks like this is the issue: [google/sanitizers#1716](https://github.com/google/sanitizers/issues/1716) There is a suggested workaround but not sure if I want to do it.
> 
> Edit: Looks like we applied that workaround for ASAN at some point: #13150 . Then reverted it: #13266 as it is not needed for ASAN anymore.

it's not that ASAN doesn't need it, but GH uses `vm.mmap_rnd_bits=28` by default, so we revert it.

sundb commented on tsan.supp:
```diff
@@ -0,0 +1,6 @@
+# collect_stacktrace_data() calls backtrace() from a signal handler but
```
Put it in src dir? like valgrind.sup.

skaslev replied:
Done.

sundb commented on tests/unit/info.tcl:
```diff
@@ -349,7 +349,9 @@ start_server {tags {"info" "external:skip"}} {
             if {$::verbose} { puts "instantaneous metrics instantaneous_eventloop_duration_usec: $value" }
             assert_morethan $value 0
-            assert_lessthan $value [expr $retries*22000] ;# default hz is 10, so duration < 1000 / 10, allow some tolerance
+            if {!$::tsan} {
```
what about tag these tests as asan:skip, then skip then when ::tsan is true?

skaslev replied:
Great idea, added tsan:skip tag to skip tests that fail due to timing issues with thread sanitizer.

sundb commented on tests/unit/client-eviction.tcl:
```diff
@@ -479,8 +479,8 @@ start_server {} {
-        wait_for_condition 200 100 {
-            [llength [regexp -all -inline {name=client} [r client list]]] == $client_count / 2
+        wait_for_condition 20 100 {
+            [llength [regexp -all -inline {name=client} [r client list]]] <= $client_count / 2
```
why change this?

skaslev replied:
Two parts here. 
The change to the first line is now reverted (will push shortly). It was there initially because a lot of timing related tests were failing with thread sanitizer. The problem was that tests run 16 redis clients by default, which taken with `--config io-threads 4`, oversaturates many machines and the timings tests expect don't align. Instead we have `--clients 1` on the yaml CI workflow file for thread san runs now.
The second change in necessary because redis with `io-threads 4` config might evict slightly more clients (prob. because of the additional memory overhead) the test now expects that half or more of the active clients might have been evicted.

sundb replied:
Perhaps we can determine whether running with io-thread. If it is, we will use the new verification and then add a TODO in the subsequent PR to fix it.
like
```
        wait_for_condition 200 100 {
            ([lindex [r config get io-threads] 1] == 1) ?
                ([llength [regexp -all -inline {name=client} [r client list]]] <= $client_count / 2) :
                ([llength [regexp -all -inline {name=client} [r client list]]] == $client_count / 2)
        } else {
            fail "Failed to evict clients"
        }
```

skaslev replied:
Done.

sundb commented on tests/test_helper.tcl:
```diff
@@ -542,8 +548,8 @@ proc print_help_screen {} {
         "--quiet            Don't show individual tests."
-        "--single <unit>    Just execute the specified unit (see next option). This option can be repeated."
         "--verbose          Increases verbosity."
+        "--single <unit>    Just execute the specified unit (see next option). This option can be repeated."
```
please revert this change.

skaslev replied:
Done.

sundb commented on tests/unit/memefficiency.tcl:
```diff
@@ -931,7 +931,7 @@ run_solo {defrag} {
     }
 
-    start_cluster 1 0 {tags {"defrag external:skip cluster"} overrides {appendonly yes auto-aof-rewrite-percentage 0 save "" loglevel notice}} {
+    start_cluster 1 0 {tags {"defrag external:skip external:tsan cluster"} overrides {appendonly yes auto-aof-rewrite-percentage 0 save "" loglevel notice}} {
```
should `external:tsan` be `tsan:skip`?

skaslev replied:
Yep, thank you :)

tezc commented on .github/workflows/daily.yml:
```diff
@@ -676,7 +676,7 @@ jobs:
           ref: ${{ env.GITHUB_HEAD_REF }}
       - name: make
-        run: make SANITIZER=undefined REDIS_CFLAGS='-DREDIS_TEST -Werror' SKIP_VEC_SETS=yes LUA_DEBUG=yes # we (ab)use this flow to also check Lua C API violations
+        run: make SANITIZER=undefined REDIS_CFLAGS='-DREDIS_TEST -Werror -DDEBUG_ASSERTIONS' SKIP_VEC_SETS=yes LUA_DEBUG=yes # we (ab)use this flow to also check Lua C API violations
```
why do we have this change? it seems unrelated to tsan, no? if this is unnecessary, please revert not to pollute blame log.

skaslev replied:
Done

tezc commented on src/debug.c:
```diff
@@ -2583,6 +2583,7 @@ void serverLogHexDump(int level, char *descr, void *value, size_t len) {
 #include <sys/time.h>
 
 void sigalrmSignalHandler(int sig, siginfo_t *info, void *secret) {
+    int save_errno = errno;
```
could you tell us why do we need this change? Maybe we should a comment to describe why we need it. e.g. We need it for TSAN, otherwise it gives false-positive errors.

skaslev replied:
Tsan reports `WARNING: ThreadSanitizer: signal handler spoils errno` so we save and restore `errno`.
I added a comment as you suggested.

tezc commented on src/tsan.sup:
```diff
@@ -0,0 +1,6 @@
+# glibc 2.39 it does through a call to _dl_map_object_deps().
+signal:collect_stacktrace_data
+signal:printCrashReport
+race:malloc_mutex_trylock_final
```
We managed to avoid suppression files for other sanitizers, I wonder if we can get rid of this file as well. First two can be solved with using `REDIS_NO_SANITIZE("thread")` in the code but we can't do anything for the last one right? It is inside jemalloc if I'm not mistaken

tezc replied:
I see you build with `USE_JEMALLOC=no`, is this function inside libc malloc?

skaslev replied:
You're right. We no longer need the first two, since stacktrace producing tests are now disabled in tsan runs.

For the last one in jemalloc, tsan reports it even with explicit `__attribute__((no_sanitize("thread")))` on `malloc_mutex_trylock_final()` so I don't see how to get away with `tsan.sup`:

```
WARNING: ThreadSanitizer: data race (pid=2232559)
  Atomic read of size 1 at 0x7d7e4f603928 by main thread (mutexes: write M0, write M1):
    #0 pthread_mutex_trylock ../../../../src/libsanitizer/tsan/tsan_interceptors_posix.cpp:1354 (libtsan.so.2+0x5832a) (BuildId: 38097064631f7912bd33117a9c83d08b42e15571)
    #1 malloc_mutex_trylock_final include/jemalloc/internal/mutex.h:158 (redis-server+0x3c68a9) (BuildId: 3b576a58153c6c4f34bf5c56cf129de1065b4f80)
    #2 malloc_mutex_lock include/jemalloc/internal/mutex.h:217 (redis-server+0x3c68a9)
    #3 je_arena_stats_merge src/arena.c:154 (redis-server+0x3c68a9)
    #4 cronUpdateMemoryStats /home/skaslev/src/redis/src/server.c:1363 (redis-server+0xcc3c7) (BuildId: 3b576a58153c6c4f34bf5c56cf129de1065b4f80)
    #5 cronUpdateMemoryStats /home/skaslev/src/redis/src/server.c:1348 (redis-server+0xcc3c7)
    #6 serverCron /home/skaslev/src/redis/src/server.c:1475 (redis-server+0xdaedf) (BuildId: 3b576a58153c6c4f34bf5c56cf129de1065b4f80)
    #7 serverCron /home/skaslev/src/redis/src/server.c:1408 (redis-server+0xdaedf)
    #8 processTimeEvents /home/skaslev/src/redis/src/ae.c:330 (redis-server+0xa13e6) (BuildId: 3b576a58153c6c4f34bf5c56cf129de1065b4f80)
    #9 aeProcessEvents /home/skaslev/src/redis/src/ae.c:465 (redis-server+0xa7057) (BuildId: 3b576a58153c6c4f34bf5c56cf129de1065b4f80)
    #10 aeProcessEvents /home/skaslev/src/redis/src/ae.c:360 (redis-server+0xa7057)
    #11 aeMain /home/skaslev/src/redis/src/ae.c:495 (redis-server+0xa7057)
    #12 main /home/skaslev/src/redis/src/server.c:7624 (redis-server+0x8e0f6) (BuildId: 3b576a58153c6c4f34bf5c56cf129de1065b4f80)

  Previous write of size 1 at 0x7d7e4f603928 by thread T2 (mutexes: write M2):
    #0 pthread_mutex_init ../../../../src/libsanitizer/tsan/tsan_interceptors_posix.cpp:1315 (libtsan.so.2+0x594cd) (BuildId: 38097064631f7912bd33117a9c83d08b42e15571)
    #1 je_malloc_mutex_init src/mutex.c:171 (redis-server+0x4216cd) (BuildId: 3b576a58153c6c4f34bf5c56cf129de1065b4f80)
    #2 bioProcessBackgroundJobs /home/skaslev/src/redis/src/bio.c:359 (redis-server+0x23e352) (BuildId: 3b576a58153c6c4f34bf5c56cf129de1065b4f80)

  Mutex M0 (0x5696643bd520) created at:
    #0 pthread_mutex_lock ../../../../src/libsanitizer/tsan/tsan_interceptors_posix.cpp:1341 (libtsan.so.2+0x59a13) (BuildId: 38097064631f7912bd33117a9c83d08b42e15571)
    #1 moduleInitModulesSystem /home/skaslev/src/redis/src/module.c:12214 (redis-server+0x2b398c) (BuildId: 3b576a58153c6c4f34bf5c56cf129de1065b4f80)
    #2 main /home/skaslev/src/redis/src/server.c:7363 (redis-server+0x8dcfe) (BuildId: 3b576a58153c6c4f34bf5c56cf129de1065b4f80)

  Mutex M1 (0x5696645d4e20) created at:
    #0 pthread_mutex_init ../../../../src/libsanitizer/tsan/tsan_interceptors_posix.cpp:1315 (libtsan.so.2+0x594cd) (BuildId: 38097064631f7912bd33117a9c83d08b42e15571)
    #1 je_malloc_mutex_init src/mutex.c:171 (redis-server+0x4216cd) (BuildId: 3b576a58153c6c4f34bf5c56cf129de1065b4f80)

  Mutex M2 (0x5696643d4240) created at:
    #0 pthread_mutex_init ../../../../src/libsanitizer/tsan/tsan_interceptors_posix.cpp:1315 (libtsan.so.2+0x594cd) (BuildId: 38097064631f7912bd33117a9c83d08b42e15571)
    #1 je_malloc_mutex_init src/mutex.c:171 (redis-server+0x4216cd) (BuildId: 3b576a58153c6c4f34bf5c56cf129de1065b4f80)

  Thread T2 'bio_aof' (tid=2232564, running) created by main thread at:
    #0 pthread_create ../../../../src/libsanitizer/tsan/tsan_interceptors_posix.cpp:1022 (libtsan.so.2+0x5ac1a) (BuildId: 38097064631f7912bd33117a9c83d08b42e15571)
    #1 bioInit /home/skaslev/src/redis/src/bio.c:173 (redis-server+0x23df7c) (BuildId: 3b576a58153c6c4f34bf5c56cf129de1065b4f80)
    #2 InitServerLast /home/skaslev/src/redis/src/server.c:3031 (redis-server+0xd230b) (BuildId: 3b576a58153c6c4f34bf5c56cf129de1065b4f80)
    #3 main /home/skaslev/src/redis/src/server.c:7574 (redis-server+0x8dfbf) (BuildId: 3b576a58153c6c4f34bf5c56cf129de1065b4f80)

SUMMARY: ThreadSanitizer: data race include/jemalloc/internal/mutex.h:158 in malloc_mutex_trylock_final
```

tezc replied:
wait, you build redis with `USE_JEMALLOC=no` in the daily.yml. The above one is a build with jemalloc. Are we going to use jemalloc or not? Is `USE_JEMALLOC=no` needed now?

skaslev replied:
I've missed the `USE_JEMALLOC=no` when testing locally.

We can either drop `USE_JEMALLOC=no` for tsan runs in CI or keep it and also keep the `malloc_mutex_trylock_final()` in `src/tsan.sup`.

Also, I double checked and the the first two are also needed in `tsan.supp` unless we disable more tests. Marking the functions with REDIS_NO_SANITIZE doesn't have effect.

skaslev replied:
Dropped `USE_JEMALLOC=no` from daily CI thread sanitizer runs.
Couldn't figure out to go away with the suppression file.

tezc replied:
let's drop USE_JEMALLOC=no and keep the file.

tezc commented on src/server.h:
```diff
@@ -1757,10 +1757,10 @@ struct redisServer {
-    volatile sig_atomic_t shutdown_asap; /* Shutdown ordered by signal handler. */
-    volatile sig_atomic_t crashing;      /* Server is crashing report. */
+    redisAtomic int shutdown_asap; /* Shutdown ordered by signal handler. */
+    redisAtomic int crashing;      /* Server is crashing report. */
```
I wonder if we are getting this right. Documentation mentions `volatile sig_atomic_t` is guaranteed to work in a signal handler. I understand the reason is need for compiler and CPU barriers. So, with redisAtomic, we will also get the same properties. Though, now, the access to these variables should use atomicSet()/atomicGet() functions.

ShooterIT replied:
i think sanitizer doesn't report errors after changing to `redisAtomic`,  the reason is that the implementation of `redisAtomic` is `_Atomic` for new compiler version(newer than 4.9), so `++`, `--` and `=` of them are `memory_order_seq_cst` semantics even we don't use `atomicSet()/atomicGet()`

skaslev replied:
I was also under the impression that `volatile sig_atomic_t` is enough nevertheless thread sanitizer still flags it as problematic.
@ShooterIT is right, `_Atomic` shouldn't need `atomicSet()/atomicGet()`
I'm open to suggestions here; redisAtomic works and I didn't come up with any better alternatives.

tezc replied:
My question is what happens on an older compiler without `_Atomic`? Because `_Atomic`  was introduced in C11. Now, read/write to these variables are not guaranteed to be atomic and there is no visibility guarantee without atomicGet() / atomicSet(), no? Maybe I'm missing something..

ShooterIT replied:
@tezc you are right, i just said why this code can work, as you said, the sanitizer may report errors if we use gcc 4.8. we should use `atomicSet()/atomicGet()` if we want to this change compatible

> I was also under the impression that volatile sig_atomic_t is enough nevertheless thread sanitizer still flags it as problematic.

I agree volatile sig_atomic_t is enough is enough, maybe  it is a false positive, sanitizer can't distinguish the signal hander and thread

tezc replied:
I'm concerned about regular build rather than TSAN false reports. Code may not work correctly without atomicGet() / atomicSet() on an older compiler.

skaslev replied:
I think the story behind "`volatile sig_atomic_t` is enough" is that in single threaded setup, the signal handler is called while the main thread is suspended. But with many threads the kernel can decide calling the the handler on any other thread (modulo `pthread_sigmask()`) while the main thread is still running.
@ShooterIT suggested using `pthread_sigmask()` to ensure that the signal handler is always called on the main thread. This is very cool since it avoids use of atomic instructions whatsoever. Unfortunately I couldn't make this approach work, because I have no control over the sigmask of threads created by modules.
Instead, added `atomicGet/Set()` to the appropriate places so that we support compilers without C11's _Atomic, as @tezc suggested.
Can you please take another look?

tezc replied:
It's a bit ugly to use atomicGet()/atomicSet() instead of direct access but there is no better alternative if I understand correctly. Also, initially I thought this was a false positive but looks like a potential issue in our signal handler. Could you please mention in the top comment why we should not use `sig_atomic_t`, it is an interesting detail anyway :)

tezc replied:
@ShooterIT any other ideas to avoid this?

ShooterIT replied:
so `volatile sig_atomic_t` always has an issue whatever we enable io threading or not since we always have many bg threads, maybe multiple io threads makes this issue easy to appear?

If there is a real issue, i don't have another idea but using `atomicGet()/atomicSet()` that is very annoying, as @skaslev said `pthread_sigmask` can't work well with module threads.

BTW, maybe in the next big version, we can consider not supporting old compliers that don't support atomic variable.

tezc replied:
yeah, looks like the issue has been there forever but probably very unlikely to happen. Okay, so we have atomicGet()/atomicSet() now and we should be safe.

tezc commented on .github/workflows/daily.yml:
```diff
@@ -697,6 +697,53 @@ jobs:
+        run: HAS_TSAN=1 TSAN_OPTIONS="detect_deadlocks=0,suppressions=src/tsan.sup" CFLAGS='-Werror' ./runtest-moduleapi --clients 1 --config io-threads 4 --accurate --verbose --dump-logs ${{github.event.inputs.test_args}}
+      - name: sentinel tests
+        if: true && !contains(github.event.inputs.skiptests, 'sentinel')
+        run: HAS_TSAN=1 TSAN_OPTIONS="detect_deadlocks=0,suppressions=src/tsan.sup" ./runtest-sentinel ${{github.event.inputs.cluster_test_args}}
```
Couple of minor details, 

1- See we pass SANITIZER options here: https://github.com/redis/redis/blob/22ebb06eb3d76575434674c7913635fea5393774/tests/support/server.tcl#L297

Maybe we can embed TSAN_OPTIONS there. 

2- Regarding `HAS_TSAN`, maybe it will be more in line if we use a flag to the test suite, e.g. --tsan similar to --valgrind. I don't have a strong opinion but it also avoids surprises if there is a leaked `HAS_TSAN` env variable. @sundb maybe you can comment if you have a preference.

sundb replied:
@tezc good idea.

skaslev replied:
Done: moved TSAN_OPTIONS setting where ASAN/MSAN_OPTIONS is set and moved to using --tsan flag instead of HAS_TSAN env variable.

Thank for suggesting. Please take another look.

sundb commented on src/tsan.sup:
```diff
@@ -0,0 +1,7 @@
+signal:collect_stacktrace_data
+signal:printCrashReport
+# TODO Investigate this race in jemalloc
+race:malloc_mutex_trylock_final
```
maybe related to: https://github.com/jemalloc/jemalloc/issues/2621

skaslev replied:
Added reference to this issue in `tsan.sup`

sundb commented on .github/workflows/daily.yml:
```diff
@@ -737,13 +737,10 @@ jobs:
       - name: sentinel tests
         if: true && !contains(github.event.inputs.skiptests, 'sentinel')
-        run: ./runtest-sentinel --tsan ${{github.event.inputs.cluster_test_args}}
+        run: ./runtest-sentinel --config io-threads 4 ${{github.event.inputs.cluster_test_args}}
```
does it mean that cluster and sentinel don't run thread sanitizer?
please note that we already have a similar CI in test-ubuntu-io-threads.
EDIT: the cluster and the sentinel doesn't need to ignore tests, so we don't need `--tsan`.

tezc commented:
@ShooterIT? Is there a real race condition triggered by this test?

@skaslev correct me if I'm wrong, if a test is not using backtrace, we skipped it because of timing issue. So, I assume this test should be skipped due to timing issue but as far as I can see, there is no heavy memory usage in this test. Maybe, this test was failing due to a real issue, not by a false-positive?

skaslev commented:
Ok, I remember seeing "test {client tracking don't cause eviction feedback loop}" fail under tsan. Bad judgement on my part for considering it timing related, but I haven't seen any races reported by tsan for it though.
In general, all non producing backtrace test should be uskipped under tsan after unflaking them at some point.
